### PR TITLE
strformat.fmt now supports non-literal const strings

### PR DESCRIPTION
**Pull request by timotheecour**
* strformat.fmt now supports non-literal const strings
* undeprecate callsite:
  * it's useful in this PR
  * see also rationale i gave in https://github.com/nim-lang/Nim/pull/11864#issuecomment-517392813

## note 1
I'm not supporting `&` with non-constant string literals in this PR because it's hitting https://github.com/nim-lang/Nim/issues/18278 and https://github.com/nim-lang/Nim/issues/18275, and because fmt can aways be used instead of & (see details in PR)

## note 2
after this PR, we get clean sigmatch errors (or allow overload resolution to favor other overloads) rather than error inside the instantiation eg in cases like this:

```nim
when defined case11:
  import std/strformat
  proc main =
    let c1 = '{'
    let c2 = '}'
    let b = "abc".fmt(c1, c2)
    echo b
  main()
```

this also improves the siganture for & to use the `string{lit}` instead of a hiddent assumption

### CODE REVIEW

**Comment by Araq** [comment]
> undeprecate callsite: it's useful in this PR

Callsite is both expensive for compile-times (it requires a copyTree operation for every call expression) and against Nim's design. I won't accept undeprecation.

**Comment by timotheecour** [comment]
PTAL and see updated PR description

> Callsite is both expensive for compile-times (it requires a copyTree operation for every call expression) and against Nim's design. I won't accept undeprecation.

i've reverted the undeprecation (sorry for the triple negative), and have a plan to replace it without loosing anything, see https://github.com/nim-lang/RFCs/issues/387

also, seems to me `&` is entirely redundant with fmt as i noted in the runnableExamples (unless i'm missing something?) and also worse because & is already used for binary append, and as you noted `"git clone {url}"` is easier to confuse with `&"git clone {url}"` than `fmt"git clone {url}"`

but I haven't marked `&` as deprecated, instead marked fmt as preferable since it's better to have some guidelines than let multiple styles persist when one has objective advantages to the other.

**Comment by Araq** [comment]
> but I haven't marked & as deprecated, instead marked fmt as preferable since it's better to have some guidelines than let multiple styles persist when one has objective advantages to the other.

The `&` syntax is the preferred one though as ``fmt"{x}\n"`` doesn't do the right thing.

**Comment by timotheecour** [comment]
> The & syntax is the preferred one though as fmt"{x}\n" doesn't do the right thing.

that's a really moot point given that with fmt you can do both raw and regular literals via fmt"\n" vs "\n".fmt, unlike with &, and raw vs regular strings are well documented, IMO.

And all the other arguments point in the direction favoring `fmt`:
* & is non-standard notation, and confusing with binary &, and hard to search, whereas fmt is self-documenting and uniquely named
* fmt is more flexible; besides raw+regular, it allows custom {} chars
* fmt now allows non-literal const strings, whereas supporting this for & runs into compiler bugs #18278 and #18275, which, while probably fixable, are a direct result of name clash with binary &

**Comment by Araq** [comment]
> that's a really moot point given that with fmt you can do both raw and regular literals via fmt"\n" vs "\n".fmt, unlike with &, and raw vs regular strings are well documented, IMO.

Many people complained about it, well documented or not. And to be fair I can understand it's rather error-prone.

**Comment by timotheecour** [comment]
@araq to reply to this:
> plus almost every usage of strformat that I've seen was an abuse of this feature
> and for me personally it's unlikely to get so good that I want to use it

Here's a perfect use case for `fmt` where alternatives are inferior:
## fmt
```nim
fmt"{dir}/t{num}.nim and {x}, {y}: {z}."

# possible in future work (i implemented it in my own code):
fmt("$dir/t$num.nim and $x, $y: $z.", '$')
# or:
fmts"$dir/t$num.nim and $x, $y: $z."
```

## concat (upcoming PR):
concat is great, but for this use case fmt is better:
```nim
concat(dir, "/t", num, ".nim", ext, " and ", x, ", ", y, ": ", z, ".")
```

## `& + $` append
this is less efficient than fmt or concat because of temporaries ($x vs addInt etc), even with term rewriting macros;
it's arguably less readable, most people would agree with that (can you spot the error? this is so common)
```nim
dir & "t/" & $num & ".nim" & " and " & $x & ", " & $y & ": " $z & "."
```

## strutils.%:
arguably less readable as fmt since it's not clear which variable is which:
less efficient than fmt and concat for same reason (temporaries):
```nim
"$#/t$#.nim and $#, $#: $#." % [dir, $num, $x, $y, $z]
```

more readable that $# but so verbose people usually don't bother writing code like that
even less efficient IIUC because of the temporaries (and there are more of them)
```nim
"$dir/t$num.nim and $x, $y: $z." % ["dir", dir, "num", $num, "x", $x, "y", $y, "z", $z]
```

## interp
proposed in https://github.com/nim-lang/Nim/issues/18218#issuecomment-861713652 and implemented but no PR yet because of too many open discussions at the moment:

more efficient than strutils.% and `& + $` append (no temporaries)
useful in cases where format string is known only at RT (or if you want this to work inside a template, although fmt can be made to work too, https://github.com/nim-lang/Nim/issues/10977 is fixable) 
```nim
interp "{dir}/t{num}.nim and {x}, {y}: {z}.", (dir, num, x, y, z)
interp "$dir/t$num.nim and $x, $y: $z.", (dir, num, x, y, z), esc = '$'
```

## in summary:
formatting is one of the most common tasks, it makes sense to have more than 1 tool for the job to avoid "1 size fits none"; I'd argue you only need 3:

* fmt is arguably the better fit in examples like the above where you have a lot of static strings in the pattern (this is what i meant in the chat) as opposed to just variables

* concat is arguably the better fit in cases where you're mostly taking in variables that are not static strings, eg:
```nim
fmt"{a}{b}{c}" vs concat(a, b, c)

# or with a sep:
fmt"{x[0]}, {n.sym.name}, {t.len}, {d}"
vs
concat(x[0], n.sym.name, t.len, d, sep = ", ")
```

* interp is better when the pattern is known at RT instead of CT (in those cases fmt and concat can't be used)

**Comment by Araq** [comment]
> concat is great, but for this use case fmt is better

Can't argue with that. You're right. :-)

However, for me this would be acceptable too:

```nim

const
  comma = ", "
  colon = ": "
  dot = "."

concat(dir, "/t", num, ".nim", ext, " and ", x, comma, y, colon, z, dot)
```

File: lib/pure/strformat.nim [hunk 1]
```diff
@@ -573,15 +573,14 @@ template formatValue(result: var string; value: char; specifier: string) =
 template formatValue(result: var string; value: cstring; specifier: string) =
   result.add value
 
-proc strformatImpl(pattern: NimNode; openChar, closeChar: char): NimNode =
-  if pattern.kind notin {nnkStrLit..nnkTripleStrLit}:
-    error "string formatting (fmt(), &) only works with string literals", pattern
+proc strformatImpl(f: string; openChar, closeChar: char): NimNode =
+  let info = callsite()
+  # pending https://github.com/nim-lang/RFCs/issues/387, use `callTree`
```
Comment (Araq) [inline lib/pure/strformat.nim]: Can you find a way to do this for now without callsite and without RFC 387 ?

Comment (timotheecour) [reply lib/pure/strformat.nim]: last commit removes callsite (but I still think RFC 387 is a good thing to implement regardless), here's the semantic difference:
```nim
when defined case1:
  import std/strformat
  let x = @[3, 4]
  let a = fmt"foo {nonexistent}"

when defined case2:
  import std/strformat
  let x = @[3, 4]
  let a = fmt"foo {x * }"
```
case1: no difference
case2:
before PR:
```
stack trace: (most recent call last)
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/strformat.nim(673, 53) fmt
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/strformat.nim(639, 18) strformatImpl
/Users/timothee/git_clone/nim/timn/tests/nim/all/t12419.nim(5, 1) template/generic instantiation from here
/Users/timothee/git_clone/nim/timn/tests/nim/all/t12419.nim(13, 14) Error: could not parse `x * `.
/Users/timothee/git_clone/nim/Nim_devel/lib/core/macros.nim(530, 5) Error: expression expected, but found '[EOF]'
    let a = fmt"foo {x * }"
```

after PR (after the commit that removes callsite):
```
stack trace: (most recent call last)
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/strformat.nim(680, 16) fmt
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/strformat.nim(634, 16) strformatImpl
/Users/timothee/git_clone/nim/timn/tests/nim/all/t12419.nim(5, 1) template/generic instantiation from here
/Users/timothee/git_clone/nim/timn/tests/nim/all/t12419.nim(13, 14) template/generic instantiation of `fmt` from here
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/strformat.nim(634, 16) Error: could not parse `x * ` in `foo {x * }`.
/Users/timothee/git_clone/nim/Nim_prs/lib/core/macros.nim(531, 5) Error: expression expected, but found '[EOF]'
            error("could not parse `$#` in `$#`.\n$#" % [subexpr, f, e.msg])
```

i'd argue it's better after PR because it shows more context in stacktrace and all the information is still there.

File: lib/pure/strformat.nim [hunk 2]
```diff
@@ -667,19 +660,39 @@ proc strformatImpl(pattern: NimNode; openChar, closeChar: char): NimNode =
...
+    assert "`x`".fmt('`', '`') == "7"
+  strformatImpl(pattern, openChar, closeChar)
+
+template fmt*(pattern: static string): untyped =
+  ## Alias for `fmt(pattern, '{', '}')`.
+  fmt(pattern, '{', '}')
+
+macro `&`*(pattern: string{lit}): string =
+  ## `&pattern` is the same as `pattern.fmt`; prefer `fmt` since it's more
+  ## flexible and readable (`&` can be confused with binary append operator).
```
Comment (Araq) [inline lib/pure/strformat.nim]:
```suggestion
  ## `&pattern` is the same as `pattern.fmt`.
  ## For a specification of the `&` macro, see the module level documentation.
```
